working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+otherwise the result of calling "toString".
+
+**COMMENTARY(leafp):** *This needs a bit of work.  If we keep the dynamic check
+  for "struct-ness", then we need do deal with the case that the user defines a
+  debutToString thing with the wrong type*.
+
+**COMMENTARY(leafp):** *If we keep this, we may wish to specify that compilers
+  may choose to make it a link time error to have an invocation of debugToString
+  in the program outside of asserts, and other debutToString methods*.
+
+
+#### copyWith
+
+If no copyWith method is defined in or inherited by the struct, then a copyWith
+method will be defined with named parameters for every entry in the primary
+constructor.  For every entry, if the field name is not private, then the
+parameter name is the field name.  If the field name is private, then the
+parameter name is the field name with all leading `_`s removed.
+
+The body of the method calls the generated primary constructor.  For every
+argument which is explicitly passed to the copyWith method invocation, the
+corresponding parameter is passed on as the argument to the corresponding
+parameter of the generated primary constructor.  For every argument which is not
+passed to the copyWith method invocation, the value of the corresponding field
+from the current instance is passed on as the argument to the corresponding
+parameter of the generated primary constructor.
+
+**COMMENTARY(leafp):** *As with constructors, this method cannot be generated as
+  a strict de-sugaring, since the ability to detect whether or not an argument
+  has been passed is not available in Dart.  Implementations already support
+  this for default values, so it is likely not problematic to implement*.
+
+
+#### Additional generated members
+
+We may wish to also define additional generated members.  For example, we may
+wish to have `parse` and `unParse` methods.  TODO(leafp): consider fleshing
+something out here.
+
+
+#### Restrictions
+
+It is an error to override the "runtimeType" method of a struct.
+
+It is an error to override the "noSuchMethod" method of a struct.
+
+### Const structs
+
+TODO(leafp): This should work, write out the details.
+
+## Extension Structs
+
+Extension structs are restrictions of the core struct feature, designed to
+support wrapper-less views on an object.  This section describes how extension
+structs look and behave, largely by example.  In many cases, alternatives or
+extensions to the core proposal are described in line.
+
+## Extension structs by example
+
+Extension structs are targeted at relatively niche uses where you wish to define
+a set of statically dispatched methods layered on top of an underlying
+representation, without introducing a wrapper object.  A canonical use case
+driving this design is to be able define a Dart typed interface for methods on a
+Javascript object, providing a wrapperless interoperation capability.
+
+
+```dart
+extension struct Window(JSObject o) {
```

Comment:
Can you illustrate what a non-default constructor might look like for extension structs? I ask because it might be tempting for a user to write something like the following:

```
extension struct Window(JSObject o) {
  Window.cons() => Window(js_util.callConstructor(...));
}
```

to avoid duplicating logic to construct `o` every time.

Reply:
My initial pitch would be that you just use the normal constructor syntax.  So your example would look like:

```dart
extension struct Window(JSObject o) {
  Window.cons() : o = Window(js_util.callConstructor(...));
}
```

Factory constructors are also available.  

If that's not enough, we could consider allowing the field to be initialized in the constructor body, with the restriction that `this` cannot be accessed before the field is definitely assigned.  That's more of an implementation lift though.

Reply:
Oh, this is great! External constructors might be a bit more complicated if we intend to support that (perhaps they just desugar to something like the above), but this should cover any cases we have today for interop constructors (and more).

> allowing the field to be initialized in the constructor body

This might be useful in the case where you have a complicated initializer, but I supposed you can always have a separate initialization method that you can call to instantiate `Window` e.g. `: o = Window(initializer())`.

Reply:
You can't access `this`, so `intializer` can't be an instance method.  But otherwise yes.  And you do have factory constructors.  I think I should probably change this to make the generated constructor always be named `._`, so then you could do:

```
extension struct Window(JSObject o) {
  factory Window.cons() {
     var o = Window(...);
      return Window._(o);
  }
}
```